### ISSUE TITLE
vim + ALE + analyser: Some errors/fixes not detected until file is saved

### ISSUE THREAD

**Issue by dorfsmay** [rust-analyzer/rust-analyzer#6346]
When using vim + ALE + rust-anlyser, some errors are highlighted when the buffer is changed, but others only when the file is saved.

Using this example:
```
fn main() {
  let x = 3;
  x += 4;
  println!("{}", x); 
}
```

When loading the file, it shows the error "cannot assign twice to immutable variable `x`" on line 3, which is expected.

If I add a second t to "let" on the second line, and immediately get a syntax error. On the other hand, if I add the word `mut` on line 2, the error about a second assignment on line 3 does not go away until I save the file.

Again, if I delete the word 'mut', I don't get an error, not until I save the file.

Using vim + ALE with javascript, I have never run into a similar issue, so this seems specific to rust-analyser. Is this a rust-analyser issue? Is this by design? Is there a way to make it work?

**Comment by kiljacken** [comment]
Some diagnostics in rust-analyzer are provided by running `cargo check`. As cargo check doesn't have an API for providing a VFS we can only run it for saved files, and thus only trigger it when files are saved.

**Comment by jonas-schievink** [comment]
Yeah this is currently working as intended, so closing.
